src/custom_collections.jl

```diff
@@ -343,4 +344,48 @@ end
 @inline Base.setindex!(v::DiscardVector, value, i::Int) = nothing
 @inline Base.size(v::DiscardVector) = (v.length,)
 
+
+const MatrixBlock{T} = SubArray{T,2,Array{T, 2},Tuple{UnitRange{Int},UnitRange{Int}},false}
+
+"""
+$(TYPEDEF)
+
+`SegmentedBlockDiagonalMatrix` is an `AbstractMatrix` backed by a plain Matrix which
+additionally stores a sequence of views into the diagonal blocks of the parent matrix. This
+type is useful for storing and updating block-diagonal matrices whose block contents
+may change but whose overall structure is fixed, such as configuration derivative <-> velocity
+jacobians.
+"""
+struct SegmentedBlockDiagonalMatrix{T} <: AbstractMatrix{T}
+    parent::Matrix{T}
+    blocks::Vector{MatrixBlock{T}}
+
+    function SegmentedBlockDiagonalMatrix{T}(parent::AbstractMatrix{T}, block_indices) where T
```

Comment:
This accepts `AbstractMatrix{T}`, but the field type is `Matrix{T}`. So if you call the constructor with a non-`Matrix` `AbstractMatrix`, the argument will be converted to a `Matrix`, meaning that the parent of each of the `blocks` is no longer the same as the `parent` field.

Should probably just parameterize this type on the type of `parent`, and make `MatrixBlock` parametric as well.

Reply:
fixed